File: frontend/src/types/analysis.ts
// Types mirroring the backend Phase 4 analysis schemas.

export type Severity = "info" | "low" | "medium" | "high" | "critical";
export type Confidence = "low" | "medium" | "high";

export interface PRMeta {
  owner: string;
  repo: string;
  number: number;
  title: string;
  state: string;
  author: string | null;
  body: string | null;
  html_url: string;
  base_branch: string | null;
  head_branch: string | null;
  created_at: string | null;
  updated_at: string | null;
}

export interface Finding {
  id: string;
  category: string;
  severity: Severity;
  title: string;
  file: string | null;
  line_range: [number, number] | null;
  what_changed: string;
  why_it_matters: string;
  evidence: string[];
  confidence: Confidence;
}

export interface FunctionalArea {
  name: string;
  files: string[];
  detected_by: string;
}

export interface ScopeDriftSignal {
  has_drift: boolean;
  stated_areas: string[];
  actual_areas: string[];
  unexpected_areas: string[];
  reason: string;
}

export interface ChangeSignal {
  files_changed: number;
  lines_added: number;
  lines_removed: number;
  total_changes: number;
  size_class: string;
  category_counts: Record<string, number>;
  functional_areas: FunctionalArea[];
  scope_drift: ScopeDriftSignal;
  findings: Finding[];
}

export interface CriticalArea {
  name: string;
  kind: string;
  files: string[];
  evidence: string[];
  severity: Severity;
}

export interface CriticalAreaSignal {
  areas: CriticalArea[];
  findings: Finding[];
}

export interface SecurityCheckResult {
  check_type: string;
  status: "pass" | "warn" | "fail";
  summary: string;
  findings: Finding[];
}

export interface SecuritySignal {
  findings: Finding[];
  checks: SecurityCheckResult[];
  total_checks: number;
  passed_checks: number;
  failed_checks: number;
}

export interface QualitySignal {
  findings: Finding[];
}

export interface TestCoverageIndication {
  changed_source_files: number;
  changed_test_files: number;
  source_files_with_tests: number;
  source_files_without_tests: number;
  missing_test_files: string[];
  has_reliable_coverage_report: boolean;
  coverage_note: string;
}

export interface TestSignal {
  coverage: TestCoverageIndication;
  findings: Finding[];
}

export interface AnalysisResult {
  phase: string;
  meta: PRMeta;
  change: ChangeSignal;
  critical: CriticalAreaSignal;
  security: SecuritySignal;
  quality: QualitySignal;
  tests: TestSignal;
  all_findings: Finding[];
  stats: Record<string, number>;
  project_context?: ProjectContext;
}

export interface ProjectContext {
  name: string;
  description: string;
  language: string;
  tech_stack: string[];
  key_directories: string[];
  readme_excerpt: string;
  package_info: Record<string, string>;
}

export interface FindingExplanation {
  finding_id: string;
  explanation: string;
  is_inference: boolean;
}

export interface ScopeDriftReasoning {
  assessment: string;
  is_inference: boolean;
}

export interface AIInsights {
  pr_intent: string;
  functional_area_interpretation: string;
  scope_drift_reasoning: ScopeDriftReasoning | null;
  finding_explanations: FindingExplanation[];
  risk_reasoning: string;
  is_ai_generated: boolean;
  provider: string;
  disclaimer: string;
}

export interface RiskWeights {
  change: number;
  scope: number;
  critical: number;
  security: number;
  quality: number;
  tests: number;
}

export interface RiskDimensionScore {
  dimension: string;
  raw_score: number;
  weight: number;
  weighted_score: number;
  contributing_factors: string[];
}

export interface RiskResult {
  total_score: number;
  level: string;
  dimension_scores: RiskDimensionScore[];
  top_contributors: string[];
  weights: RiskWeights;
  disclaimer: string;
}

export interface PriorityItem {
  rank: number;
  file: string;
  score: number;
  severity: Severity;
  reason: string;
  supporting_finding_ids: string[];
}

export interface ReviewPriority {
  items: PriorityItem[];
  top_concerns: string[];
}

export interface Phase4Result {
  phase: string;
  meta: PRMeta;
  analysis: AnalysisResult;
  ai_insights: AIInsights;
  risk: RiskResult;
  priority: ReviewPriority;
  stats: Record<string, number | string>;
}

// Phase 1 types (for diff)
export interface ChangedFile {
  filename: string;
  status: string;
  additions: number;
  deletions: number;
  changes: number;
  patch: string | null;
  previous_filename: string | null;
}

export interface DiffSummary {
  total_additions: number;
  total_deletions: number;
  total_changes: number;
  files_changed: number;
}

export interface AnalyzeResponse {
  phase: string;
  meta: PRMeta;
  diff_summary: DiffSummary;
  files: ChangedFile[];
  raw_diff: string | null;
}

File: backend/models/analysis.py
"""Pydantic schemas for the Phase 3 analysis result.

The AnalysisResult is the structured, evidence-backed output of the
deterministic analysis engine. It is consumed by Phase 4 (AI + Risk + Priority).
"""
from __future__ import annotations

from enum import Enum
from typing import Dict, List, Optional, Tuple

from pydantic import BaseModel, Field

from models.schemas import PRMeta


class Severity(str, Enum):
    INFO = "info"
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class Confidence(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"


class Finding(BaseModel):
    """A single evidence-backed analysis finding.

    Every finding attempts to answer WHAT / WHERE / WHY / IMPACT.
    """

    id: str = Field(..., description="Stable identifier for this finding")
    category: str = Field(
        ..., description="change | critical | security | quality | tests"
    )
    severity: Severity = Severity.INFO
    title: str
    file: Optional[str] = None
    line_range: Optional[Tuple[int, int]] = None
    what_changed: str = ""
    why_it_matters: str = ""
    evidence: List[str] = Field(default_factory=list)
    confidence: Confidence = Confidence.MEDIUM


class FunctionalArea(BaseModel):
    name: str
    files: List[str] = Field(default_factory=list)
    detected_by: str = Field(
        default="path", description="path | keyword | content"
    )


class ScopeDriftSignal(BaseModel):
    """Signal indicating the PR may be broader than its stated intent."""

    has_drift: bool = False
    stated_areas: List[str] = Field(
        default_factory=list,
        description="Functional areas inferred from PR title/body",
    )
    actual_areas: List[str] = Field(
        default_factory=list,
        description="Functional areas inferred from changed files",
    )
    unexpected_areas: List[str] = Field(
        default_factory=list,
        description="Actual areas not mentioned in the PR title/body",
    )
    reason: str = ""


class ChangeSignal(BaseModel):
    files_changed: int = 0
    lines_added: int = 0
    lines_removed: int = 0
    total_changes: int = 0
    size_class: str = Field(
        default="small", description="small | medium | large | xlarge"
    )
    category_counts: Dict[str, int] = Field(default_factory=dict)
    functional_areas: List[FunctionalArea] = Field(default_factory=list)
    scope_drift: ScopeDriftSignal = Field(default_factory=ScopeDriftSignal)
    findings: List[Finding] = Field(default_factory=list)


class CriticalArea(BaseModel):
    name: str
    kind: str = Field(
        ..., description="authentication | authorization | payments | sensitive-data | infrastructure | business-critical"
    )
    files: List[str] = Field(default_factory=list)
    evidence: List[str] = Field(default_factory=list)
    severity: Severity = Severity.HIGH


class CriticalAreaSignal(BaseModel):
    areas: List[CriticalArea] = Field(default_factory=list)
    findings: List[Finding] = Field(default_factory=list)


class SecurityCheckResult(BaseModel):
    """Result of a single security check category."""

    check_type: str = Field(..., description="secret-scan | injection | xss | csrf | path-traversal | ssrf | command-injection | sensitive-file | dependency")
    status: str = Field(..., description="pass | warn | fail")
    summary: str = ""
    findings: List[Finding] = Field(default_factory=list)


class SecuritySignal(BaseModel):
    findings: List[Finding] = Field(default_factory=list)
    checks: List[SecurityCheckResult] = Field(default_factory=list)
    total_checks: int = 0
    passed_checks: int = 0
    failed_checks: int = 0


class QualitySignal(BaseModel):
    findings: List[Finding] = Field(default_factory=list)


class TestCoverageIndication(BaseModel):
    """A coverage *indication*, never an invented exact percentage."""

    # Prevent pytest from trying to collect this Pydantic model as a test class.
    __test__ = False

    changed_source_files: int = 0
    changed_test_files: int = 0
    source_files_with_tests: int = 0
    source_files_without_tests: int = 0
    missing_test_files: List[str] = Field(default_factory=list)
    has_reliable_coverage_report: bool = False
    coverage_note: str = Field(
        default="Test signal based on changed-file and naming-convention heuristics."
    )


class TestSignal(BaseModel):
    # Prevent pytest from trying to collect this Pydantic model as a test class.
    __test__ = False

    coverage: TestCoverageIndication = Field(default_factory=TestCoverageIndication)
    findings: List[Finding] = Field(default_factory=list)


class AnalysisResult(BaseModel):
    """Top-level structured analysis output (Phase 3)."""

    phase: str = "phase-3-analysis-engine"
    meta: PRMeta
    change: ChangeSignal = Field(default_factory=ChangeSignal)
    critical: CriticalAreaSignal = Field(default_factory=CriticalAreaSignal)
    security: SecuritySignal = Field(default_factory=SecuritySignal)
    quality: QualitySignal = Field(default_factory=QualitySignal)
    tests: TestSignal = Field(default_factory=TestSignal)
    all_findings: List[Finding] = Field(default_factory=list)
    stats: Dict[str, int] = Field(default_factory=dict)
